rework `std.Build` usage/integration

mlugg opened the pull request:
This commit effectively rewrites `build.zig`, `build/TranslateC.zig` (now `build/Translator.zig`), and the example build scripts.

Instead of being its own step, `Translator` is a helper type which simply wraps a `Build.Step.Run` and an associated `*Build.Module`. This is not only simpler, but also necessary, because custom steps are going to be removed from `std.Build` at some point. Many of the same APIs as before are available, but also several more. For instance, you can call `Translator.linkLibrary` to both link a library (as a `Step.Compile`) to the generated `*Build.Module`, *and* expose its installed headers to `translate-c`. In general, any method of `*Build.Module` which can add an include path or similar is replicated on `Translator`.

The `build.zig` has been improved and simplified. The old script was quite overcomplicated, and also relied on some hacks and even technically on some Illegal Behavior. Much of this complexity arose from the fact that the script was trying to make `zig build test` run the translate-c tests in *all* optimization modes. This is unnecessary effort; it's much simpler to just run `zig build` a few separate times from the CI workflow.

The `test-examples` step was previously disabled due to a Zig compiler bug with the recursive imports (`build.zig` depends on `examples/x/build.zig` which depends on `build.zig` again). While the exact current output is a compiler bug, it's actually not possible for recursive package dependencies to work; you would end up infinitely recursing through `fn build` calls. So, the `test-examples` step is removed. It is replaced with a new `examples/build.zig`, which just aggregates all of the examples so you can run `zig build all` in that directory to test them all. The CI workflow now runs this command too.

Lastly, the examples are refactored to use the new APIs, be a bit clearer, and to properly test their output (`import_header` by actually running the test it builds, `compile_c` by checking the captured stdout of both executables after running them).

Oh, and I added a new example, `use_static_lib`, demonstrating the ability of the new API to completely wrap a C library (by translating its headers, but also including its source in the generated module) in a concise way.

mlugg committed:
rework `std.Build` usage/integration

This commit effectively rewrites `build.zig`, `build/TranslateC.zig`
(now `build/Translator.zig`), and the example build scripts.

Instead of being its own step, `Translator` is a helper type which
simply wraps a `Build.Step.Run` and an associated `*Build.Module`. This
is not only simpler, but also necessary, because custom steps are going
to be removed from `std.Build` at some point. Many of the same APIs as
before are available, but also several more. For instance, you can call
`Translator.linkLibrary` to both link a library (as a `Step.Compile`) to
the generated `*Build.Module`, *and* expose its installed headers to
`translate-c`. In general, any method of `*Build.Module` which can add
an include path or similar is replicated on `Translator`.

The `build.zig` has been improved and simplified. The old script was
quite overcomplicated, and also relied on some hacks and even
technically on some Illegal Behavior. Much of this complexity arose from
the fact that the script was trying to make `zig build test` run the
translate-c tests in *all* optimization modes. This is unnecessary
effort; it's much simpler to just run `zig build` a few separate times
from the CI workflow.

The `test-examples` step was previously disabled due to a Zig compiler
bug with the recursive imports (`build.zig` depends on
`examples/x/build.zig` which depends on `build.zig` again). While the
exact current output is a compiler bug, it's actually not possible for
recursive package dependencies to work; you would end up infinitely
recursing through `fn build` calls. So, the `test-examples` step is
removed. It is replaced with a new `examples/build.zig`, which just
aggregates all of the examples so you can run `zig build all` in that
directory to test them all. The CI workflow now runs this command too.

Lastly, the examples are refactored to use the new APIs, be a bit
clearer, and to properly test their output (`import_header` by actually
running the test it builds, `compile_c` by checking the captured stdout
of both executables after running them).

Oh, and I added a new example, `use_static_lib`, demonstrating the
ability of the new API to completely wrap a C library (by translating
its headers, but also including its source in the generated module) in a
concise way.

mlugg commented:
I just need to fix up some system integration bits; Linux is working because it's the one where Aro's toolchain works natively.

There's also a function in the build system abstraction -- currently unused, hence why Linux CI passed -- which is using a currently-nonexistent function `std.Build.Step.ConfigHeader.getOutputDir`. That's added in ziglang/zig#24205

mlugg committed:
Update Aro dependency

mlugg committed:
examples: handle Windows CRT stdout line endings

On Windows, the CRT (libc) will by default have stdio streams open in
"text" mode, so will convert "\n" into "\r\n" in output. We could
change the file mode in the examples behind a `#ifdef _WIN32`, but it's
a bit easier to just handle the difference in the build script checks.

mlugg committed:
build: use Zig library directories for cross-translation

Only use Aro's native Linux toolchain on a native Linux target.

mlugg committed:
build.zig.zon: bump `minimum_zig_version`

This version is necessary because:

* It introduced `std.Build.Step.ConfigHeader.getOutputDir`
* It is Aro's new minimum

mlugg committed:
ci: test against `minimum_zig_version`

mlugg/setup-zig v2 will do this by default if the version isn't
specified in the manifest, and it seems like a better idea here. We're
keeping `minimum_zig_version` up-to-date anyway; this change decreases
maintenance burder if and when we pin to full Zig releases, and it means
that if there is a breaking change upstream, our CI doesn't break.

ehaas commented:
Running `zig build` on this branch on an aarch64 mac seems to get stuck in "Configure"

```
ehaas@Mac ~ % zig version
0.15.0-dev.847+850655f06
```
```
ehaas@Mac translate-c % git rev-parse --short HEAD
c39ac9b
```

```
ehaas@Mac ~ % uname -a
Darwin Mac 24.5.0 Darwin Kernel Version 24.5.0: Tue Apr 22 19:53:27 PDT 2025; root:xnu-11417.121.6~2/RELEASE_ARM64_T6041 arm64
```

```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x000000019ca25498 libsystem_kernel.dylib`poll + 8
    frame #1: 0x0000000100587d34 build`posix.poll + 116
    frame #2: 0x0000000100660014 build`io.Poller(process.Child.collectOutput__enum_101809).pollPosix + 120
    frame #3: 0x0000000100660650 build`io.Poller(process.Child.collectOutput__enum_101809).poll + 60
    frame #4: 0x0000000100660ab0 build`process.Child.collectOutput + 320
    frame #5: 0x0000000100610454 build`process.Child.run + 416
    frame #6: 0x00000001005b681c build`zig.system.darwin.getSdk + 496
    frame #7: 0x000000010055ccd4 build`zig.LibCInstallation.findNative + 364
    frame #8: 0x000000010055c48c build`zig.LibCDirs.detect + 1000
    frame #9: 0x000000010053868c build`build.Translator.initInner + 1688
    frame #10: 0x0000000100538cd8 build`test.cases.lowerCases + 1000
    frame #11: 0x0000000100521fa8 build`build.build + 5968
    frame #12: 0x00000001005112d0 build`Build.runBuild__anon_21347 + 24
    frame #13: 0x000000010050a0a4 build`build_runner.main + 15340
    frame #14: 0x0000000100514248 build`main + 612
    frame #15: 0x000000019c6beb98 dyld`start + 6076
```

ehaas commented:
Update: actually it was just really slow:

```
ehaas@Mac translate-c % time zig build
zig build  2.57s user 47.60s system 94% cpu 53.367 total
```

Vexu reviewed (approved):
A bit unfortunate that the examples can't be run from the root but at least they can be run from one place now. I'll add a note to the README when I update it.

mlugg committed:
build: cache `std.zig.LibCDirs.detect` result

This code isn't great, but it works around a major performance issue
when building for native macOS. We might be able to remove this once
Aro's toolchains advance to the point where we only need to use this
code path for cross-compilation, since that detection should be fast.

mlugg committed:
build: introduce `-Dtest-cross-targets`

This new `zig build` option causes the cases to be translated not only
for the `-Dtarget` (typically `native`) target, but also (unless they
have an explicit target override) a list of 5 cross targets:

* `x86_64-linux`
* `x86_64-windows`
* `aarch64-linux`
* `aarch64-macos`
* `arm-linux`

The CI workflow is updated to test these cross targets (in Debug mode).

Almost every existing case was able to pass on these cross targets
without any changes. Aro was missing support for varargs on ARM (see
Vexu/arocc#869) which caused any test importing `stdio.h` to fail on
`arm-linux`, and one of the run cases needed an `#ifdef` added to avoid
assuming `__int128`, but that's all!

mlugg commented:
@ehaas commit 72acbbd introduced some slightly hacky caching for the libc dir detection (to avoid it being repeated for every test case, i.e. hundreds of times!). Would you be able to re-test on the latest version of this PR and confirm whether the "Configure" performance issue has been resolved?

ehaas commented:
Performance looks good!

```
ehaas@Mac translate-c % time zig build
zig build  0.07s user 0.55s system 95% cpu 0.650 total
```

mlugg committed:
ci: don't test every build mode

It's overkill. Just Debug and ReleaseFast will be plenty.

Also, don't do a ReleaseSafe build; again, there just isn't any point.
